bolts/models/dann.py

```diff
@@ -0,0 +1,195 @@
+        self._loss_clf_fn = nn.BCEWithLogitsLoss()
+
+        self.test_acc_s = torchmetrics.Accuracy()
+        self.test_acc_y = torchmetrics.Accuracy()
+        self.train_acc_s = torchmetrics.Accuracy()
+        self.train_acc_y = torchmetrics.Accuracy()
+        self.val_acc_s = torchmetrics.Accuracy()
+        self.val_acc_y = torchmetrics.Accuracy()
+
+    def _inference_epoch_end(
+        self, output_results: List[Dict[str, Tensor]], stage: Stage
+    ) -> Dict[str, Tensor]:
+        all_y = torch.cat([_r["y"] for _r in output_results], 0)
+        all_s = torch.cat([_r["s"] for _r in output_results], 0)
+        all_preds = torch.cat([_r["preds"] for _r in output_results], 0)
+
+        dt = em.DataTuple(
+            x=pd.DataFrame(
+                torch.rand_like(all_s, dtype=float).detach().cpu().numpy(), columns=["x0"]
+            ),
+            s=pd.DataFrame(all_s.detach().cpu().numpy(), columns=["s"]),
+            y=pd.DataFrame(all_y.detach().cpu().numpy(), columns=["y"]),
+        )
+
+        results = em.run_metrics(
+            predictions=em.Prediction(hard=pd.Series(all_preds.detach().cpu().numpy())),
+            actual=dt,
+            metrics=[em.Accuracy(), em.RenyiCorrelation(), em.Yanovich()],
+            per_sens_metrics=[em.Accuracy(), em.ProbPos(), em.TPR()],
+        )
+
+        tm_acc_s = self.val_acc_s if stage == "val" else self.test_acc_s
+        tm_acc_y = self.val_acc_y if stage == "val" else self.test_acc_y
+        acc_s = tm_acc_s.compute().item()
+        acc_y = tm_acc_y.compute().item()
+        results_dict = {f"{stage}/acc_s": acc_s, f"{stage}/acc_y": acc_y}
+        results_dict.update({f"{stage}/{k}": v for k, v in results.items()})
+        return results_dict
+
+    def _inference_step(self, batch: DataBatch, stage: Stage) -> Dict[str, Tensor]:
+        _s, _y = self.forward(batch.x)
+        target_s = batch.s.view(-1, 1).float()
+        loss_adv = self._loss_adv_fn(_s, target_s)
+        target_y = batch.y.view(-1, 1).float()
+        loss_clf = self._loss_clf_fn(_y, target_y)
+        loss = loss_adv + loss_clf
+        tm_acc_s = self.val_acc_s if stage == "val" else self.test_acc_s
+        tm_acc_y = self.val_acc_y if stage == "val" else self.test_acc_y
+
+        target_y = batch.y.view(-1, 1).long()
+        y_acc = tm_acc_y(_y >= 0, target_y)
+
+        target_s = batch.s.view(-1, 1).long()
+        s_acc = tm_acc_s(_s >= 0, target_s)
+        self.log_dict(
+            {
+                f"{stage}/loss": loss.item(),
+                f"{stage}/loss_adv": loss_adv.item(),
+                f"{stage}/loss_clf": loss_clf.item(),
+                f"{stage}/acc_s": s_acc,
+                f"{stage}/acc_y": y_acc,
+            }
+        )
+        return {"y": batch.y, "s": batch.s, "preds": _y.sigmoid().round().squeeze(-1)}
+
+    def reset_parameters(self) -> None:
+        """Reset the models."""
+        self.adv.apply(self._maybe_reset_parameters)
```

Comment:
apply is recursive - it's sufficient just to 'apply' `_maybe_reset_parameters` to self